### PR TITLE
KAFKA-4913: creating a window store with one segment throws division by zero error

### PR DESCRIPTION
**Pull request by dguy**
Allow a window store with a single segment.

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang @enothereska @xvrl @mjsax

**Comment by guozhangwang** [comment]
> If there is only one segment then set the cleanup.policy to compact.

What's the motivation behind this (with one window segment the changelog topic may still have lots of log segments)?

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 1]
```diff
@@ -111,6 +111,12 @@ public StateStoreSupplier build() {
 
                                     @Override
                                     public PersistentKeyValueFactory<K, V> windowed(final long windowSize, final long retentionPeriod, final int numSegments, final boolean retainDuplicates) {
+                                        if (numSegments < 1) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java:114]: Since we are already checking this in `RocksDBWindowStoreSupplier`, do we still need the same checking here?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 2]
```diff
@@ -45,15 +45,24 @@
...
+        if (numSegments < 1) {
+            throw new IllegalArgumentException("numSegments must be positive");
+        }
+        if (retentionPeriod < 1) {
+            throw new IllegalArgumentException("retentionPeriod must be positive");
+        }
         this.name = name;
         this.numSegments = numSegments;
-        this.segmentInterval = Math.max(retentionPeriod / (numSegments - 1), MIN_SEGMENT_INTERVAL);
+        this.segmentInterval = numSegments == 1 ? Long.MAX_VALUE : Math.max(retentionPeriod / (numSegments - 1), MIN_SEGMENT_INTERVAL);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java:56]: Another thought while reading this line: maybe we should just forbid setting num.segments to 1 since otherwise we will not able to expire any windows at all, like the way we did here?

**Comment by dguy** [comment]
> What's the motivation behind this (with one window segment the changelog topic may still have lots of log segments)?

If we use delete then we will be deleting data that belongs in the segment. Or maybe it doesn't matter as it should obey the retention period anyway. Yeah i think that is right! Ok i'll change that back! Thanks

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 2 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java:56]: I considered that too. I don't feel strongly about it either way. @xvrl you raised the issue - thoughts?

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java (hunk 1 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/Stores.java:114]: Yes, i think it is good practice to do it here as it is thrown as early as possible. We also need to have it `RocksDBWindowStoreSupplier` as there is nothing to stop users from just constructing that class.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 3]
```diff
@@ -111,6 +111,12 @@ public StateStoreSupplier build() {
...
+                                            throw new IllegalArgumentException("numSegments must be positive");
+                                        }
+                                        if (retentionPeriod < 1) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java:117]: Should this not be an even stronger check: `retentionPeriod < windowSize` ? (some below)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java:56]: We actually do that for a min of 2 segments in `Windows` already: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java#L89-L92

So how is is even possible to specify a single segment in the first place? Do you create the store manually?

Thus, I think we should keep a min segment size of two. (Or be very clear that retention time will not have any effect if only a single segment is used)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java [hunk 4]
```diff
@@ -45,15 +45,24 @@
...
         // Create a date formatter. Formatted timestamps are used as segment name suffixes
         this.formatter = new SimpleDateFormat("yyyyMMddHHmm");
         this.formatter.setTimeZone(new SimpleTimeZone(0, "UTC"));
     }
 
     long segmentId(long timestamp) {
+        if (numSegments == 1) {
+            return 1;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java:64]: `return 0;` -> we always starting counting at zero ;P

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 2 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java:56]: @mjsax you can use `Stores.create(..)...` or you can use the `RocksDBStoreSupplier` . But yeah we could make it a min of 2 unless we think there is valid reason why you'd want a window store with a single segment.

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/Stores.java:117]: @dguy ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStoreSupplier.java [hunk 5]
```diff
@@ -46,6 +46,12 @@ public RocksDBWindowStoreSupplier(String name, long retentionPeriod, int numSegm
 
     public RocksDBWindowStoreSupplier(String name, long retentionPeriod, int numSegments, boolean retainDuplicates, Serde<K> keySerde, Serde<V> valueSerde, Time time, long windowSize, boolean logged, Map<String, String> logConfig, boolean enableCaching) {
         super(name, keySerde, valueSerde, time, logged, logConfig);
+        if (numSegments < 1) {
+            throw new IllegalArgumentException("numSegments must be positive");
+        }
+        if (retentionPeriod < 1) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWindowStoreSupplier.java:52]: As above.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 4 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java:64]: @dguy ? Why do we number segments starting from `1` ?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java [hunk 6]
```diff
@@ -854,7 +854,7 @@ public Object apply(Object value1, Object value2) {
                         return null;
                     }
                 },
-                JoinWindows.of(0)
+                JoinWindows.of(1)
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java:857]: Why this change?

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java (hunk 4 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java:64]: @mjsax - i don't remember!

**Comment by dguy** [comment]
@guozhangwang i'm not particularly bothered with this patch. Maybe we should do it differently, i.e., only allow >=2 segments for a Window store. I haven't seen a valid argument for having a single segment. WDYT?

**Comment by guozhangwang** [comment]
> @guozhangwang i'm not particularly bothered with this patch. Maybe we should do it differently, i.e., only allow >=2 segments for a Window store. I haven't seen a valid argument for having a single segment. WDYT?

I agree, we can just forbid setting one segment.

Also there are a few comments from @mjsax that seem not addressed yet?
